modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1325 @@
+}
+
+pragma "no doc"
+private extern proc sizeof(e): size_t;
+
+extern "struct event_base" record event_base {};
+extern "struct event" record event {};
+
+extern const EV_TIMEOUT:c_short;
+extern const EV_READ:c_short;
+extern const EV_WRITE:c_short;
+extern const EV_SIGNAL:c_short;
+extern const EV_PERSIST:c_short;
+extern const EV_ET:c_short;
+extern const EV_FINALIZE:c_short;
+extern const EV_CLOSED:c_short;
+extern const EVLOOP_NO_EXIT_ON_EMPTY:c_int;
+
+private extern proc event_base_new():c_ptr(event_base);
+private extern proc event_base_dispatch(base: c_ptr(event_base)):c_int;
+private extern proc event_base_loop(base: c_ptr(event_base), flags: c_int):c_int;
+private extern proc event_base_free(base: c_ptr(event_base));
+private extern proc event_base_got_break(base: c_ptr(event_base)):c_int;
+private extern proc event_base_loopbreak(base: c_ptr(event_base)):c_int;
+private extern proc event_new(base: c_ptr(event_base), fd: c_int, events: c_short, callback: c_fn_ptr, callback_arg: c_void_ptr): c_ptr(event);
+private extern proc event_add(ev: c_ptr(event), timeout: c_ptr(timeval)):c_int;
+private extern proc event_del(ev: c_ptr(event)):c_int;
+private extern proc event_free(ev: c_ptr(event)):c_int;
+private extern proc event_remove_timer(ev: c_ptr(event)):c_int;
+private extern proc evutil_make_socket_nonblocking(fd: c_int):c_int;
+private extern proc libevent_global_shutdown();
+
+extern type pthread_t;
+extern type pthread_attr_t;
+private extern proc pthread_create(thread: c_ptr(pthread_t), const attr: c_ptr(pthread_attr_t), start_routine: c_fn_ptr, arg: c_void_ptr): c_int;
+private extern proc pthread_join(thread: pthread_t, retval: c_ptr(c_void_ptr)): c_int;
+private extern proc evthread_use_pthreads();
+
+evthread_use_pthreads();
+var event_loop_base = event_base_new();
+
+proc dispatchLoop():c_void_ptr throws {
+  if event_loop_base == nil {
+    throw new Error("event loop wasn't initialized");
+  }
+  if event_base_got_break(event_loop_base) == 1 {
+    return nil;
+  }
+  var x = event_base_loop(event_loop_base, EVLOOP_NO_EXIT_ON_EMPTY);
+  if x != 0 {
+    throw new Error("event loop wasn't initialized");
+  }
+  return nil;
+}
+
+var event_loop_thread:pthread_t;
+pthread_create(c_ptrTo(event_loop_thread), nil:c_ptr(pthread_attr_t), c_ptrTo(dispatchLoop), nil);
+
+proc syncRWTCallback(fd: c_int, event: c_short, arg: c_void_ptr) {
+  var syncVariablePtr = arg: c_ptr(sync int);
+  syncVariablePtr.deref().writeEF(event);
+}
+
+proc deinit() {
+  event_base_loopbreak(event_loop_base);
+  pthread_join(event_loop_thread, nil:c_ptr(c_void_ptr));
+  event_base_free(event_loop_base);
+  libevent_global_shutdown();
+}
+
+/*
+  A record holding reference to a tcp socket
+  bound and listening for connections.
+*/
+record tcpListener {
+  /*
+    File Descriptor Associated with instance
+  */
+  var socketFd: int(32) = -1;
+
+  pragma "no doc"
+  proc init(socketFd: c_int) {
+    this.socketFd = socketFd;
+  }
```

Comment:
Your tcpListener also needs a `deinit` which probably should close the socket.
It's important to still have a `close` method since `close` can fail (because there is no way of getting the returned error out of a `deinit`).